# 4.8 Reverse Digits

Write a program which takes an integer and returns the integer corresponding to the digits of the input written in reverse order. For example, the reverse of 42 is 24, and the reverse of -314 is -413

"42"
iterate through half the string and swap the digits
iterate half the array with 2 pointers and swap each pointer's value


In [18]:
def reverse(x):
    result, x_remaining = 0, abs(x)
    while x_remaining:        
        result = result * 10 + x_remaining % 10
        print(result)
        x_remaining //= 10
#        print(x_remaining)
    return -result if x < 0 else result
reverse(401)

1
10
104


104

# Note to Self:

## Floor-Divide and Assign(//=)
Performs floor-division on the values on either side. then assigns it to the expression on the left

In [10]:
x = 20
x //= 5
print(x)

4


# Analyzing code

While the input number is not zero
    we take the resulting number multiply it by zero and add the mod 10 of the input number
    update the input number to be divided and floored by 10

since we're dealing with decimal numbers, for every digit that exists the first digit will be multiplied by 10^n-1
n being the number of digits 

x = 401
result = 0
x_remaining = +(positive) 401 

while(x_remaining)
    result = 0 * 10 + 401 % 10
    x_remaining //= 10 // dividing and flooring = 40

result = 1 
x_remaining = 40

while(x_remaining):
    result = 1 * 10 + 40 % 10 = 1*10 + 0 = 10
    x_remaining //= 10 // dividing and flooring = 4

result = 1 
x_remaining = 4


while (x_remaining):
    result = 10 * 10 + 4 % 10 = 100 + 4 = 104
       x_remaining //=10 # x_remaining becomes zero

loop breaks and our result is the reversed digit
104
    

In [14]:
x = 1132
print(x % 10) # retrieving the first digit of 1132 by mod 10, remainder of 1132/10 = 2
x //= 10
print(x)

2
113
